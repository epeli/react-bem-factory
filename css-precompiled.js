module.exports = require("./lib/css-precompiled");
