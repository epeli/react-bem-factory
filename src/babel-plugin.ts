import * as BabelTypes from "@babel/types";
import { Visitor, NodePath } from "@babel/traverse";
import { SourceMapGenerator } from "source-map";
import convert from "convert-source-map";

type CallValue = BabelTypes.CallExpression["arguments"][0];

interface BabelFile {
    opts: {
        generatorOpts: any;
    };
    code: string;
    path: NodePath;
}

export interface PluginOptions {
    opts?: {
        target?: string;
        runtime?: string;
    };
    file: BabelFile;
}

export interface Babel {
    types: typeof BabelTypes;
}

function getGeneratorOpts(
    file: BabelFile,
): { sourceFileName: string; sourceRoot: string } {
    return file.opts.generatorOpts ? file.opts.generatorOpts : file.opts;
}

export function makeSourceMapGenerator(file: BabelFile) {
    const generatorOpts = getGeneratorOpts(file);
    const filename = generatorOpts.sourceFileName;
    const generator = new SourceMapGenerator({
        file: filename,
        sourceRoot: generatorOpts.sourceRoot,
    });

    generator.setSourceContent(filename, file.code);
    return generator;
}

export function getSourceMap(
    offset: {
        line: number;
        column: number;
    },
    file: BabelFile,
): string {
    const generator = makeSourceMapGenerator(file);
    const generatorOpts = getGeneratorOpts(file);
    if (generatorOpts.sourceFileName) {
        generator.addMapping({
            generated: {
                line: 1,
                column: 0,
            },
            source: generatorOpts.sourceFileName,
            original: offset,
        });
        return convert.fromObject(generator).toComment({ multiline: true });
    }
    return "";
}

export default function bemedBabelPlugin(
    babel: Babel,
): { visitor: Visitor<PluginOptions> } {
    const t = babel.types;

    /**
     * Local name of the css import from react-bemed/css if any
     */
    let name: string | null = null;

    return {
        visitor: {
            Program() {
                // Reset import name state when entering a new file
                name = null;
            },

            ImportDeclaration(path, state) {
                const opts = state.opts || {};

                const target = opts.target || "react-bemed/css";

                if (path.node.source.value !== target) {
                    return;
                }

                for (const s of path.node.specifiers) {
                    if (!t.isImportSpecifier(s)) {
                        continue;
                    }
                    if (s.imported.name === "css") {
                        name = s.local.name;
                    }
                }
            },

            TaggedTemplateExpression(path, state) {
                if (!name) {
                    return;
                }

                if (!t.isIdentifier(path.node.tag, { name })) {
                    return;
                }

                if (!path.node.loc) {
                    return;
                }

                const sourceMap = getSourceMap(path.node.loc.start, state.file);

                path.replaceWith(
                    t.callExpression(t.identifier(name), [
                        path.node.quasi,
                        t.stringLiteral(sourceMap),
                    ]),
                );
            },
        },
    };
}
