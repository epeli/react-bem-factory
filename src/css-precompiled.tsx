import { createCSSTag } from "./css-core";

function replaceCompiler(selector: string, cssString: string) {
    return cssString.replace(/__BEMED__/g, selector);
}

export const css = createCSSTag(replaceCompiler);
