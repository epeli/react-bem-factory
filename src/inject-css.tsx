import { isBrowser } from "./is-browser";
import { StyleSheet } from "@emotion/sheet";

const DELIMETER = "/*|*/";

let sheet: StyleSheet | null = null;

if (isBrowser()) {
    sheet = new StyleSheet({
        key: "bemed",
        container: document.head,
        speedy: true,
    });
}

/**
 * Use fast style injection production provided by @emotion/sheet
 */
function productionInject(id: string, css: string) {
    css.split(DELIMETER).forEach(rule => {
        if (sheet) {
            sheet.insert(rule);
        }
    });
}

const DEV_STYLE_TAGS: Record<
    string,
    { el: HTMLStyleElement; css: string } | undefined
> = {};

/**
 * In development use slower injection method that allows source maps and hot
 * reload
 */
function devInject(id: string, css: string) {
    css = css.trim();
    const existing = DEV_STYLE_TAGS[id];

    if (existing) {
        if (existing.el.innerHTML !== css) {
            console.log("react-bemed: hot update: " + id);
            existing.el.innerHTML = css;
            existing.el.dataset.ver = String(
                Number(existing.el.dataset.ver) + 1,
            );
        }
    } else {
        const style = document.createElement("style");
        style.dataset.bemed = id;
        style.dataset.ver = "1";
        style.innerHTML = css;
        document.head.appendChild(style);
        DEV_STYLE_TAGS[id] = {
            el: style,
            css,
        };
    }
}

export function injectGlobal(id: string, css: string, sourceMap: string) {
    if (!isBrowser()) {
        return;
    }

    if (process.env.NODE_ENV !== "production") {
        return devInject(id, css + "\n" + sourceMap);
    }

    return productionInject(id, css);
}
