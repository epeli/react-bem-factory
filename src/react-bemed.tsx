import { forwardRef, createElement } from "react";
import React from "react";

type BEMCSS = import("./css-core").BEMCSS;

type AnyReactComponent =
    | keyof JSX.IntrinsicElements
    | React.JSXElementConstructor<any>;

function classNameToArray(className: undefined | string | string[]) {
    return Array.isArray(className) ? className : (className || "").split(" ");
}

/**
 * Add BEM class names to any given React Component that takes a className prop
 */
function createReactBEMComponent<
    Comp extends AnyReactComponent,
    KnownMods extends Record<string, boolean | undefined>
>(opts: {
    component: Comp;
    blockClassName: string;
    knownMods: KnownMods;
    staticClassNames: string[];
    globalStaticClassNames: string[];
    modifierSeparator: string;
    css?: BEMCSS;
}) {
    type ReactProps = React.ComponentProps<Comp>;

    type FinalProps = typeof opts.knownMods extends undefined
        ? ReactProps
        : ReactProps & ModProps<KnownMods>;

    const BEMComponent = forwardRef((props: FinalProps, ref) => {
        let componentProps: Record<string, any> = {};

        /**
         * Class names passed during rendering in JSX
         */
        const runtimeClassNames = classNameToArray(props.className);

        /** Array of used BEM mods */
        const usedMods: string[] = [];

        /** css-in-js mods */
        const usedCSS: {
            className: string;
            bemCSS: BEMCSS;
        }[] = [];

        const usedModClassNames: string[] = [];

        /**
         * Custom class names from string valued mod definitions
         * Ex. from
         *
         * {
         *      mods: {
         *          foo: "my-custom-foo"
         *      }
         * }
         */
        const customModClassNames: string[] = [];

        const applyMods = (prop: string) => {
            const modType = opts.knownMods[prop];

            // Not a style mod. Just pass it as normal prop forward
            if (!modType) {
                componentProps[prop] = props[prop];
                return;
            }

            // Inactive props. Eg. mymod={false} passed
            if (!props[prop]) {
                return;
            }

            // Custom class name mod
            if (typeof modType === "string") {
                customModClassNames.push(modType);
                return;
            }

            // A BEM mod. We need to generate BEM modifier class name from this
            usedMods.push(prop);

            // The generated mod class name
            const modClassName =
                opts.blockClassName.trim() +
                opts.modifierSeparator +
                prop.trim();

            usedModClassNames.push(modClassName);

            // Class name only mod
            if (modType === true) {
                return;
            }

            // At the point it must be a css-in-js mod
            const cssMod = (modType as any) as BEMCSS;
            usedCSS.push({
                className: modClassName,
                bemCSS: cssMod,
            });
        };

        if (opts.knownMods) {
            for (const prop in props) {
                applyMods(prop);
            }
        } else {
            componentProps = props;
        }

        /**
         * Final class name to be passed to DOM
         */
        const finalClassName = [opts.blockClassName]
            .concat(usedModClassNames.sort())
            .concat(customModClassNames)
            .concat(opts.staticClassNames)
            .concat(opts.globalStaticClassNames)
            .concat(runtimeClassNames)
            .reduce(
                (acc, className) => {
                    className = className.trim();

                    if (!className) {
                        return acc;
                    }

                    // Remove duplicates
                    //
                    // Use @: prefix to avoid collision with the Object
                    // prototype properties
                    if (acc.used["@:" + className]) {
                        return acc;
                    }

                    acc.used["@:" + className] = true;
                    acc.final.push(className);

                    return acc;
                },
                {
                    final: [] as string[],
                    used: {} as Record<string, true>,
                },
            )
            .final.join(" ")
            .trim();

        const reactElement = createElement(
            opts.component,
            Object.assign({}, componentProps, {
                className: finalClassName,
                ref,
            }),
        );

        // css-in-js css for the block
        if (opts.css) {
            usedCSS.unshift({
                className: opts.blockClassName,
                bemCSS: opts.css,
            });
        }

        if (usedCSS.length > 0) {
            // This is bit weird but we do it like this because this way the
            // css-in-js does not get imported unless it is actually used
            return usedCSS[0].bemCSS.render(
                reactElement,
                usedCSS.map(css => ({
                    className: css.className,
                    cssString: css.bemCSS.cssString,
                    sourceMap: css.bemCSS.sourceMap,
                })),
            );
        }

        return reactElement;
    });

    return (BEMComponent as any) as ((props: FinalProps) => any);
}

/**
 * Convert dict of mods to boolean react props
 */
type ModProps<T> = { [P in keyof T]?: boolean };

type MethodObject = { [key: string]: (...args: any[]) => any };

/** flatten functions in an object to their return values */
type FlattenToReturnTypes<T extends MethodObject> = {
    [K in keyof T]: ReturnType<T[K]>
};

export interface BemedOptions {
    className?: string | string[];
    separators?: {
        namespace?: string;
        modifier?: string;
        element?: string;
    };
}

/**
 * Create BEMBlock component type
 */
type BEMBlock<
    Block,
    Elements extends { [key: string]: (props: any) => React.ReactNode }
> = Block & {
    className: string;
    displayName: string;
} & FlattenToReturnTypes<Elements>;

/*
type BEM<
    BEMBlockDOMElement extends ElementNames,
    BEMBlockMods extends Record<string, true | string | BEMCSS> | undefined,
    Elements extends {
        [key: string]: (
            className: string,
            isElement?: boolean,
        ) => (props: any) => React.ReactNode;
    }
> = ((
    props: BEMBlockMods extends undefined
        ? JSX.IntrinsicElements[BEMBlockDOMElement]
        : JSX.IntrinsicElements[BEMBlockDOMElement] & ModProps<BEMBlockMods>,
) => any) &
    FlattenToReturnTypes<Elements>;
*/

export function createBemed(
    prefix?: string,
    bemedOptions: BemedOptions | undefined = {},
) {
    /**
     * Define BEM Block and Elements
     */
    function defineBEMBlock<
        Elements extends {
            [key: string]: (
                className: string,
                isElement?: boolean,
            ) => (props: any) => React.ReactNode;
        },
        BEMBlockDOMElement extends AnyReactComponent = "div",
        BEMBlockMods extends
            | Record<string, true | string | BEMCSS>
            | undefined = undefined
    >(
        blockOptions:
            | {
                  as?: BEMBlockDOMElement;
                  mods?: BEMBlockMods;
                  css?: BEMCSS;
                  className?: string | string[];
                  elements?: Elements;
              }
            | undefined = {},
    ) {
        return (blockName: string, isElement?: boolean) => {
            const separators = Object.assign(
                {
                    namespace: "-",
                    modifier: "--",
                    element: "__",
                },
                bemedOptions ? bemedOptions.separators : {},
            );

            type BEMBlockProps = ModProps<BEMBlockMods>;

            let blockClassName = "";

            if (isElement) {
                blockClassName = blockName;
            } else {
                blockClassName =
                    (prefix ? prefix + separators.namespace : "") + blockName;
            }

            const globalStaticClassNames = classNameToArray(
                bemedOptions.className,
            );

            // Ensure the type is BEMBlockDOMElement and not union with "div"
            const comp: BEMBlockDOMElement = (blockOptions.as || "div") as any;

            const Block = createReactBEMComponent({
                component: comp,
                blockClassName,
                knownMods: blockOptions.mods as BEMBlockProps,
                staticClassNames: classNameToArray(blockOptions.className),
                globalStaticClassNames,
                modifierSeparator: separators.modifier,
                css: blockOptions.css,
            });

            (Block as any).displayName = `BEM(${blockClassName})`;

            const out: any = {};

            if (blockOptions.elements) {
                for (const key in blockOptions.elements) {
                    const def = blockOptions.elements[key];
                    out[key] = def(
                        blockClassName + separators.element + key,
                        true,
                    );
                }
            }

            const final = Object.assign(Block, out, {
                className: blockClassName,
            });

            return final as BEMBlock<typeof Block, Elements>;
        };
    }

    return defineBEMBlock;
}

export const bemed = createBemed("", {
    className: "bemed",
});
