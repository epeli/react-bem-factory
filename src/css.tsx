import React from "react";
import stylis from "stylis";
import { injectGlobal } from "./inject-css";
import { isBrowser } from "./is-browser";

declare const process: any;

type StyleRenderRecord = Record<string, true>;

/**
 * Record CSS strings that are rendered to DOM
 */
let BROWSER_RECORD: StyleRenderRecord | null = {};

const Context = React.createContext<StyleRenderRecord | null>(null);

export class SSRProvider extends React.Component {
    componentDidMount() {
        // Remove record after the fist browser render
        BROWSER_RECORD = null;
    }

    render() {
        return (
            <Context.Provider value={{}}>
                {this.props.children}
            </Context.Provider>
        );
    }
}

export type CSSCompiler = (className: string, css: string) => string;

const defaultCompiler = (className: string, css: string): string => {
    return stylis("." + className, css);
};

/**
 * Render given React Element in a Fragment with a style tag
 * if the given CSS chunks are not rendered before
 */
function renderWithStyleTags<T>(
    reactElement: T,
    cssChunks: {
        className: string;
        cssString: string;
    }[],
    customCompiler?: CSSCompiler,
): T {
    const cssCompiler = customCompiler || defaultCompiler;

    function render(renderRecord: StyleRenderRecord | null) {
        if (!renderRecord) {
            return reactElement;
        }

        let css = "";

        for (const chunk of cssChunks) {
            if (renderRecord[chunk.className]) {
                // Already rendered to DOM/HTML.
                continue;
            }

            renderRecord[chunk.className] = true;
            css += cssCompiler(chunk.className, chunk.cssString);
        }

        /**
         * Props for the style tag
         */
        let styleProps: any = {};

        if (process.env.NODE_ENV !== "production") {
            // For react-testing-library
            styleProps["data-testid"] = "bemed-style";
        }

        // No unrendered CSS - just return the react element
        if (!css) {
            return reactElement;
        }

        // If we have unrendered CSS render the element with a style tag
        return React.createElement(
            React.Fragment,
            null,
            React.createElement("style", styleProps, css),
            reactElement,
        ) as any;
    }

    // In browser use only a global record on the first render
    if (isBrowser()) {
        if (BROWSER_RECORD) {
            return render(BROWSER_RECORD);
        } else {
            // For subsequent render there's no need to render style tags as
            // they are injected to the HEAD
            return reactElement;
        }
    }

    // During server render get the style render record from the context so it
    // won't get mixed when multiple requests are rendered at once.
    return React.createElement(Context.Consumer, null, render) as any;
}

type Placeholders = string | number;

export function css(
    literals: TemplateStringsArray,
    ...placeholders: Placeholders[]
) {
    let cssString = "";

    for (let i = 0; i < placeholders.length; i++) {
        cssString += literals[i];
        cssString += placeholders[i];
    }

    cssString += literals[literals.length - 1];

    return {
        cssString,

        inject(className: string, compiler: CSSCompiler = defaultCompiler) {
            injectGlobal(className, compiler(className, cssString));
        },

        renderWithStyleTags,
    };
}

export type BEMCSS = ReturnType<typeof css>;
