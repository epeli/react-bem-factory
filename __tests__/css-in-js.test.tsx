import { render, cleanup, fireEvent } from "react-testing-library";
import ReactDOMServer from "react-dom/server";
import Stylis from "stylis";
import { createBemed } from "../src/react-bemed";
import React from "react";
import { css } from "../src/css";
import { css as precompiledCSS } from "../src/css-precompiled";
import { injectGlobal } from "../src/inject-css";
import {
    SSRProvider,
    _resetModuleState,
    createCSSTag,
    createClassName,
} from "../src/css-core";

jest.mock("../src/inject-css");

declare const process: any;

const mockInjectGlobal = injectGlobal as jest.Mock<typeof injectGlobal>;

const origCSSCompiler = css.compiler;
const origCSSPreCompiler = precompiledCSS.compiler;

afterEach(() => {
    cleanup();
    _resetModuleState();
    jest.resetAllMocks();
    process.env.TEST_ENV = "browser";
    process.env.NODE_ENV = "test";
    css.compiler = origCSSCompiler;
    precompiledCSS.compiler = origCSSPreCompiler;
});

test("injects style tag for blocks", () => {
    const bemed = createBemed();
    const Block = bemed({
        css: css`
            color: orange;
        `,
    })("TestBlock");

    render(<Block>test</Block>);

    expect(injectGlobal).toBeCalledTimes(1);
    expect(mockInjectGlobal.mock.calls[0][0]).toEqual("TestBlock");
    expect(mockInjectGlobal.mock.calls[0][1]).toContain("orange");
    expect(mockInjectGlobal.mock.calls[0][1]).toContain(".TestBlock");
});

test("can use variables in template literals", () => {
    const bemed = createBemed();
    const color = "orange";

    const Block = bemed({
        css: css`
            color: ${color};
        `,
    })("TestBlock");

    render(<Block>test</Block>);

    expect(injectGlobal).toBeCalledTimes(1);
    expect(mockInjectGlobal.mock.calls[0][1]).toContain("orange");
});

test("Autoprefixes during injection", () => {
    const bemed = createBemed();

    const css = createCSSTag(
        new Stylis({
            prefix: true,
        }),
    );

    const Block = bemed({
        css: css`
            @keyframes slide {
                from {
                    opacity: 0;
                }
                to {
                    opacity: 1;
                }
            }
        `,
    })("TestBlock");

    render(<Block>test</Block>);

    expect(injectGlobal).toBeCalledTimes(1);
    expect(mockInjectGlobal.mock.calls[0][1]).toContain("@-webkit-");
});

test("injects style tag for elements", () => {
    const bemed = createBemed();

    const Block = bemed({
        elements: {
            Foo: bemed({
                css: css`
                    color: red;
                `,
            }),
        },
    })("TestBlock");

    render(<Block.Foo>test</Block.Foo>);

    expect(injectGlobal).toBeCalledTimes(1);
    expect(mockInjectGlobal.mock.calls[0][0]).toEqual("TestBlock__Foo");
    expect(mockInjectGlobal.mock.calls[0][1]).toContain("red");
    expect(mockInjectGlobal.mock.calls[0][1]).toContain(".TestBlock__Foo");
});

test("injects style tag for block mods", () => {
    const bemed = createBemed();

    const Block = bemed({
        mods: {
            ding: css`
                color: orange;
            `,
        },
    })("TestBlock");

    render(<Block ding>test</Block>);

    expect(injectGlobal).toBeCalledTimes(1);
    expect(mockInjectGlobal.mock.calls[0][0]).toEqual("TestBlock--ding");
    expect(mockInjectGlobal.mock.calls[0][1]).toContain("orange");
    expect(mockInjectGlobal.mock.calls[0][1]).toContain(".TestBlock--ding");
});

test("injects style tag for block submods", () => {
    const bemed = createBemed();

    const Block = bemed({
        mods: {
            sub: {
                ding: css`
                    color: orange;
                `,
            },
        },
    })("TestBlock");

    render(<Block sub="ding">test</Block>);

    expect(injectGlobal).toBeCalledTimes(1);
    expect(mockInjectGlobal.mock.calls[0][0]).toEqual("TestBlock--sub--ding");
    expect(mockInjectGlobal.mock.calls[0][1]).toContain("orange");
    expect(mockInjectGlobal.mock.calls[0][1]).toContain(
        ".TestBlock--sub--ding",
    );
});

test("also adds the BEM class for css block mods", () => {
    const bemed = createBemed("prefix");

    const Block = bemed({
        mods: {
            ding: css`
                color: orange;
            `,
        },
    })("TestBlock");

    const rtl = render(<Block ding>test</Block>);
    const el = rtl.getByText("test");

    expect(el.className).toBe("prefix-TestBlock prefix-TestBlock--ding");
});

test("also adds the BEM class for css element mods", () => {
    const bemed = createBemed("prefix");

    const Block = bemed({
        elements: {
            Foo: bemed({
                mods: {
                    ding: css`
                        color: orange;
                    `,
                },
            }),
        },
    })("TestBlock");

    const rtl = render(<Block.Foo ding>test</Block.Foo>);
    const el = rtl.getByText("test");

    expect(el.className).toBe(
        "prefix-TestBlock__Foo prefix-TestBlock__Foo--ding",
    );
});

test("injects style tag for element mods", () => {
    const bemed = createBemed();

    const Block = bemed({
        elements: {
            Foo: bemed({
                mods: {
                    ding: css`
                        color: orange;
                    `,
                },
            }),
        },
    })("TestBlock");

    render(<Block.Foo ding>test</Block.Foo>);

    expect(injectGlobal).toBeCalledTimes(1);
    expect(mockInjectGlobal.mock.calls[0][0]).toEqual("TestBlock__Foo--ding");
    expect(mockInjectGlobal.mock.calls[0][1]).toContain("orange");
    expect(mockInjectGlobal.mock.calls[0][1]).toContain(
        ".TestBlock__Foo--ding",
    );
});

test("server renders style tags for blocks", () => {
    process.env.TEST_ENV = "node";

    const bemed = createBemed();
    const Block = bemed({
        css: css`
            color: orange;
        `,
    })("TestBlock");

    const rtl = render(
        <SSRProvider>
            <Block>test</Block>
        </SSRProvider>,
    );

    const styleTags = rtl.getAllByTestId("bemed-style");

    expect(styleTags.length).toBe(1);
    expect(styleTags[0].innerHTML).toContain("orange");
    expect(styleTags[0].innerHTML).toContain(".TestBlock");
});

test("server renders style tags for elements", () => {
    process.env.TEST_ENV = "node";

    const bemed = createBemed();
    const Block = bemed({
        elements: {
            Foo: bemed({
                css: css`
                    color: orange;
                `,
            }),
        },
    })("TestBlock");

    const rtl = render(
        <SSRProvider>
            <Block.Foo>test</Block.Foo>
        </SSRProvider>,
    );

    const styleTags = rtl.getAllByTestId("bemed-style");

    expect(styleTags.length).toBe(1);
    expect(styleTags[0].innerHTML).toContain("orange");
    expect(styleTags[0].innerHTML).toContain(".TestBlock__Foo");
});

test("does not duplicate server-rendered styles", () => {
    process.env.TEST_ENV = "node";

    const bemed = createBemed();
    const Block = bemed({
        css: css`
            color: orange;
        `,
    })("TestBlock");

    const rtl = render(
        <SSRProvider>
            <Block>test</Block>
            <Block>test</Block>
        </SSRProvider>,
    );

    const styleTags = rtl.getAllByTestId("bemed-style");

    expect(styleTags.length).toBe(1);
});

test("does not duplicate server-rendered styles within loop", () => {
    process.env.TEST_ENV = "node";

    const bemed = createBemed();
    const Block = bemed({
        css: css`
            color: orange;
        `,
    })("TestBlock");

    const rtl = render(
        <SSRProvider>
            {[1, 2, 3].map(num => (
                <Block key={num}>test</Block>
            ))}
        </SSRProvider>,
    );

    const styleTags = rtl.getAllByTestId("bemed-style");

    expect(styleTags.length).toBe(1);
});

test("server renders block mods", () => {
    process.env.TEST_ENV = "node";

    const bemed = createBemed();

    const Block = bemed({
        css: css`
            color: blue;
        `,
        mods: {
            ding: css`
                color: orange;
            `,
        },
    })("TestBlock");

    const rtl = render(
        <SSRProvider>
            <Block ding>test</Block>
        </SSRProvider>,
    );

    const styleTags = rtl.getAllByTestId("bemed-style");

    expect(styleTags.length).toBe(1);
    expect(styleTags[0].innerHTML).toContain("orange");
    expect(styleTags[0].innerHTML).toContain(".TestBlock--ding");
});

test("server renders block submods", () => {
    process.env.TEST_ENV = "node";

    const bemed = createBemed();

    const Block = bemed({
        css: css`
            color: blue;
        `,
        mods: {
            things: {
                ding: css`
                    color: orange;
                `,
            },
        },
    })("TestBlock");

    const rtl = render(
        <SSRProvider>
            <Block things="ding">test</Block>
        </SSRProvider>,
    );

    const styleTags = rtl.getAllByTestId("bemed-style");

    expect(styleTags.length).toBe(1);
    expect(styleTags[0].innerHTML).toContain("orange");
    expect(styleTags[0].innerHTML).toContain(".TestBlock--things--ding");
});

test("server renders element mods", () => {
    process.env.TEST_ENV = "node";

    const bemed = createBemed();

    const Block = bemed({
        elements: {
            Foo: bemed({
                css: css`
                    color: blue;
                `,
                mods: {
                    ding: css`
                        color: orange;
                    `,
                },
            }),
        },
    })("TestBlock");

    const rtl = render(
        <SSRProvider>
            <Block.Foo ding>test</Block.Foo>
        </SSRProvider>,
    );

    const styleTags = rtl.getAllByTestId("bemed-style");

    expect(styleTags.length).toBe(1);
    expect(styleTags[0].innerHTML).toContain("orange");
    expect(styleTags[0].innerHTML).toContain(".TestBlock__Foo--ding");
});

test("incrementally renders used css", () => {
    process.env.TEST_ENV = "node";

    const bemed = createBemed();

    const Block = bemed({
        css: css`
            color: blue;
        `,
        mods: {
            ding: css`
                color: orange;
            `,
        },
    })("TestBlock");

    const rtl = render(
        <SSRProvider>
            <div data-testid="container">
                <Block>normal</Block>
                <Block>normal dup</Block>
                <Block ding>ding</Block>
                <Block ding>ding dup</Block>
            </div>
        </SSRProvider>,
    );

    const container = rtl.getByTestId("container");

    expect(container).toMatchSnapshot();
});

test("Autoprefixes during injection", () => {
    const css = createCSSTag(
        new Stylis({
            prefix: true,
        }),
    );

    const bemed = createBemed();
    const Block = bemed({
        css: css`
            flex: 1;
        `,
    })("TestBlock");

    render(<Block>test</Block>);

    expect(injectGlobal).toBeCalledTimes(1);
    expect(mockInjectGlobal.mock.calls[0][1]).toContain("-webkit-");
});

test("server renders autoprefixed", () => {
    process.env.TEST_ENV = "node";
    const css = createCSSTag(
        new Stylis({
            prefix: true,
        }),
    );

    const bemed = createBemed();
    const Block = bemed({
        css: css`
            flex: 1;
        `,
    })("TestBlock");

    const rtl = render(
        <SSRProvider>
            <Block>test</Block>
        </SSRProvider>,
    );

    const styleTags = rtl.getAllByTestId("bemed-style");

    expect(styleTags.length).toBe(1);
    expect(styleTags[0].innerHTML).toContain("-webkit-");
});

test("can use custom stylis", () => {
    const bemed = createBemed("");

    css.compiler = new Stylis({
        prefix: false,
    });

    const Block = bemed({
        css: css`
            flex: 1;
        `,
    })("TestBlock");

    render(<Block>test</Block>);

    expect(injectGlobal).toBeCalledTimes(1);
    expect(mockInjectGlobal.mock.calls[0][1]).not.toContain("-webkit-");
});

test("can use custom css compiler for injection", () => {
    const bemed = createBemed("");

    css.compiler = () => "custom";

    const Block = bemed({
        css: css`
            flex: 1;
        `,
        mods: {
            blockMod: css`
                flex: 1;
            `,
        },
        elements: {
            Foo: bemed({
                css: css`
                    flex: 1;
                `,
                mods: {
                    elementMod: css`
                        flex: 1;
                    `,
                },
            }),
        },
    })("TestBlock");

    render(
        <div>
            <Block>test</Block>
            <Block blockMod>test</Block>
            <Block.Foo>test</Block.Foo>
            <Block.Foo elementMod>test</Block.Foo>
        </div>,
    );

    expect(injectGlobal).toBeCalledTimes(4);
    expect(mockInjectGlobal.mock.calls[0][1]).toBe("custom");
    expect(mockInjectGlobal.mock.calls[1][1]).toBe("custom");
    expect(mockInjectGlobal.mock.calls[2][1]).toBe("custom");
    expect(mockInjectGlobal.mock.calls[3][1]).toBe("custom");
});

test("can use custom css compiler in server render", () => {
    process.env.TEST_ENV = "node";
    const bemed = createBemed("");

    css.compiler = () => "custom";

    const Block = bemed({
        css: css`
            flex: 1;
        `,
        mods: {
            blockMod: css`
                flex: 1;
            `,
        },
        elements: {
            Foo: bemed({
                css: css`
                    flex: 1;
                `,
                mods: {
                    elementMod: css`
                        flex: 1;
                    `,
                },
            }),
        },
    })("TestBlock");

    const rtl = render(
        <SSRProvider>
            <Block>test</Block>
            <Block blockMod>test</Block>
            <Block.Foo>test</Block.Foo>
            <Block.Foo elementMod>test</Block.Foo>
        </SSRProvider>,
    );

    const styleTags = rtl.getAllByTestId("bemed-style");

    expect(styleTags.length).toBe(4);
    expect(styleTags.map(tag => tag.innerHTML)).toEqual([
        "custom",
        "custom",
        "custom",
        "custom",
    ]);
});

test("css can work as normal function call", () => {
    const bemed = createBemed();
    const Block = bemed({
        css: css("color: orange;", ""),
    })("TestBlock");

    render(<Block>test</Block>);

    expect(injectGlobal).toBeCalledTimes(1);
    expect(mockInjectGlobal.mock.calls[0][0]).toEqual("TestBlock");
    expect(mockInjectGlobal.mock.calls[0][1]).toEqual(
        ".TestBlock{color:orange;}",
    );
});

test("css can work as normal function call with precompiled css", () => {
    const bemed = createBemed();
    const Block = bemed({
        css: precompiledCSS("__BEMED__{color: orange;}", ""),
    })("TestBlock");

    render(<Block>test</Block>);

    expect(injectGlobal).toBeCalledTimes(1);
    expect(mockInjectGlobal.mock.calls[0][0]).toEqual("TestBlock");
    expect(mockInjectGlobal.mock.calls[0][1]).toEqual(
        ".TestBlock{color: orange;}",
    );
});

test("server render does not escape quotes", () => {
    process.env.TEST_ENV = "node";

    const bemed = createBemed();
    const Block = bemed({
        css: css`
            background-image: url("static/showell_logo_white.svg");
        `,
    })("TestBlock");

    const html = ReactDOMServer.renderToString(
        <SSRProvider>
            <Block>test</Block>
        </SSRProvider>,
    );

    expect(html).toEqual(
        '<style data-testid="bemed-style">.TestBlock{background-image:url("static/showell_logo_white.svg");}</style><div class="TestBlock">test</div>',
    );
});

test("can inject class names from createClassName()", () => {
    const bemed = createBemed();

    const FOO = createClassName(
        "foo",
        css`
            color: orange;
        `,
    );

    const Block = bemed({
        className: FOO,
    })("TestBlock");

    render(<Block>test</Block>);

    expect(injectGlobal).toBeCalledTimes(1);
    expect(mockInjectGlobal.mock.calls[0][0]).toEqual("foo");
    expect(mockInjectGlobal.mock.calls[0][1]).toContain("orange");
    expect(mockInjectGlobal.mock.calls[0][1]).toContain(".foo");
});

test("class names from createClassName() are injected before css and css mods", () => {
    const bemed = createBemed();

    const FOO = createClassName(
        "foo",
        css`
            color: orange;
        `,
    );

    const Block = bemed({
        className: FOO,
        css: css`
            color: red;
        `,
        mods: {
            mod: css`
                color: darkred;
            `,
        },
    })("TestBlock");

    render(<Block mod>test</Block>);

    expect(injectGlobal).toBeCalledTimes(3);

    // The createClassNameInjection();
    expect(mockInjectGlobal.mock.calls[0][0]).toEqual("foo");
    expect(mockInjectGlobal.mock.calls[0][1]).toContain("orange");
    expect(mockInjectGlobal.mock.calls[0][1]).toContain(".foo");

    // The main block styles
    expect(mockInjectGlobal.mock.calls[1][0]).toEqual("TestBlock");
    // Mod styles
    expect(mockInjectGlobal.mock.calls[2][0]).toEqual("TestBlock--mod");
});
