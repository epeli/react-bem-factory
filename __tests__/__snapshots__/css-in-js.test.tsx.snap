// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`incrementally renders used css 1`] = `
<div
  data-testid="container"
>
  <style
    data-testid="bemed-style"
  >
    .TestBlock{color:blue;}
  </style>
  <div
    class="TestBlock"
  >
    normal
  </div>
  <div
    class="TestBlock"
  >
    normal dup
  </div>
  <style
    data-testid="bemed-style"
  >
    .TestBlock.TestBlock--ding{color:orange;}
  </style>
  <div
    class="TestBlock TestBlock--ding"
  >
    ding
  </div>
  <div
    class="TestBlock TestBlock--ding"
  >
    ding dup
  </div>
</div>
`;
